util/ValueSetHelper.ts

```diff
@@ -0,0 +1,179 @@
+
+export interface GetValueSetCodesProps {
+  code?: string;
+  system?: string;
+}
+
+/**
+ * Helper function that takes a ValueSet url and a measure bundle and returns an
+ * array of objects that contain each code and system in the ValueSet
+ */
+function getValueSetCodes(valueSetUrl: string[], mb: fhir4.Bundle | null): GetValueSetCodesProps[] {
+  const codesAndSystems: GetValueSetCodesProps[] = [];
+  valueSetUrl.forEach(vs => {
+    const vsResource = mb?.entry?.filter(r => r.resource?.resourceType === 'ValueSet' && r.resource?.url === vs)[0]
+      .resource as fhir4.ValueSet;
+    vsResource.expansion?.contains?.forEach(c => {
+      codesAndSystems.push({ code: c.code, system: c.system });
+    });
+  });
+  return codesAndSystems;
+}
+
+/**
+ * Helper function for repeated logic in minimizeTestCaseResources that takes a primaryCodeValue,
+ * a resource, a matching DataRequirement of the same resourceType, a measureBundle, and a newResources
+ * array, checks that the primaryCodeValue exists and if the primaryCodeValue matches any of the
+ * directCodes or ValueSet codes of a DataRequirement of the same resourceType, the resource is added
+ * to the newResources array
+ */
+function checkCodesAndValueSets(
+  primaryCodeValue: fhir4.CodeableConcept,
+  matchingDRType: DataRequirementsLookupByTypeProps,
+  measureBundle: fhir4.Bundle | null,
+  resource: fhir4.BundleEntry,
+  newResources: fhir4.BundleEntry[]
+) {
+  if (primaryCodeValue) {
+    if (
+      matchingDRType.directCodes.length > 0 &&
+      matchingDRType.directCodes.find(dc => primaryCodeValue.coding?.find(c => c.code === dc.code))
+    ) {
+      newResources.push(resource);
+    } else if (matchingDRType.valueSets.length > 0) {
+      const vsCodesAndSystems = getValueSetCodes(matchingDRType.valueSets, measureBundle);
+      if (
+        vsCodesAndSystems.find(vscas =>
+          primaryCodeValue.coding?.find(c => c.code === vscas.code && c.system === vscas.system)
+        )
+      ) {
+        newResources.push(resource);
+      }
+    }
+  }
+}
+
+/**
+ * Helper function that takes in a TestCase, a measure bundle, and the
+ * current measure bundle's dataRequirements lookup by type object and returns
+ * an array of BundleEntries that are TestCase resources that are relevant to the
+ * measure bundle's data requirements
+ */
+export function minimizeTestCaseResources(
+  testCase: TestCaseInfo,
+  measureBundle: fhir4.Bundle | null,
+  drLookupByType: Record<string, DataRequirementsLookupByTypeProps>
+): fhir4.BundleEntry[] {
+  const newResources: fhir4.BundleEntry[] = [];
+  testCase.resources.forEach(r => {
+    // throw out any resources that are not in any of the dataRequirements
+    // iterate over every resource in each bundle
+    // keep Patient and MeasureReport
+    if (
+      r.resource &&
+      r.resource?.resourceType &&
+      r.resource?.resourceType !== 'MeasureReport' &&
+      r.resource?.resourceType !== 'Patient'
```

Comment:
This logic may not be necessary. It looks like `bundleToTestCase` in `util/import.ts` already takes care of filtering them out of the testCase resource list.